applications/Preprocessor/lib/ElementShapes.cpp

```diff
@@ -0,0 +1,183 @@
+ All rights reserved.
+
+ Redistribution and use in source and binary forms, with or without
+ modification, are permitted provided that the following conditions are met:
+
+ 1. Redistributions of source code must retain the above copyright notice, this
+ list of conditions and the following disclaimer.
+
+ 2. Redistributions in binary form must reproduce the above copyright notice,
+ this list of conditions and the following disclaimer in the documentation
+ and/or other materials provided with the distribution.
+
+ 3. Neither the name of the copyright holder nor the names of its contributors
+ may be used to endorse or promote products derived from this software without
+ specific prior written permission.
+
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "ElementShapes.h"
+
+namespace Preprocessor {
+
+// Forward declare some helper functions.
+// The implementation of these relies on the point and line shape, which are
+// defined only afterwards.
+namespace Detail {
+/// Helper function to quickly generate a set of boundary vertices numbered 0 to
+/// N-1
+template <std::size_t dim>
+std::vector<ElementShapePart<0, dim>> generateVertices(std::size_t numPoints);
+
+/// Helper function to generate a set of boundary edges given pairs of
+/// coordinates.
+template <std::size_t dim>
+std::vector<ElementShapePart<1, dim>> generateEdges(
+    std::initializer_list<std::vector<std::size_t>> linePoints);
+}  // namespace Detail
+
+// note when debugging new shapes: invoking the logger during static
+// initialisation can be a bit messy so you may need to use a debugger to see
+// the error message
+
+const ElementShape<0> point{};
+
+const ElementShape<1> line{Detail::generateVertices<1>(2)};
+
+const ElementShape<2> triangle{
+    // Boundary lines
+    Detail::generateEdges<2>({{0, 1}, {0, 2}, {1, 2}}),
+    // Corner vertices
+    Detail::generateVertices<2>(3)};
+const ElementShape<2> square{
+    // Boundary lines
+    Detail::generateEdges<2>({{0, 1}, {0, 2}, {1, 3}, {2, 3}}),
+    // Corner vertices
+    Detail::generateVertices<2>(4)};
+
+const ElementShape<3> tetrahedron{
+    // Faces
+    std::vector<Detail::ElementShapePart<2, 3>>{
+        Detail::ElementShapePart<2, 3>{&triangle, {0, 3, 2}},
```

Comment:
does the order matter, I guess yes? Can we add a reference to what these indices point to?

Reply:
Documented the constructor a bit further to clarify the information. The ordering matters, as it defines which triangle node matches with which node of the tetrahedron.

Reply:
Ahh so it refers to the hpgem ordering.

Reply:
I think that should be mentioned somewhere with a link to the ascii pictures

Reply:
Done